fl4health/clients/partial_weight_exchange_client.py

```diff
@@ -109,9 +107,15 @@ def set_parameters(self, parameters: NDArrays, config: Config) -> None:
                 unwound properly by the parameter exchanger
             config (Config): configuration if required to control parameter exchange.
         """
-        if not self.model_weights_initialized:
+        # Since PartialWeightExchangeClient may exchange different sets parameters
+        # across different training round, the full parameter exchanger is only used
+        # if a client is selected to participate in the first *fitting* round.
+        current_server_round = self.narrow_config_type(config, "current_server_round", int)
+        fl_fit_round = self.narrow_config_type(config, "fl_fit_round", bool)
+        if (not self.model_weights_initialized) and current_server_round == 1 and fl_fit_round:
```

Comment:
In our current design, each client uses a full parameter exchanger to pull the initial model from the server at the very beginning of FL. However, in partial exchange regimes, clients can exchange a different set of parameters with the server, so if some client is not selected to participate in the first training round and it later tries to use the full parameter exchanger to pull the partial weights, there would be a problem. The change highlighted here fixes this bug.

Reply:
> However, in partial exchange regimes, clients can exchange a different set of parameters with the server, so if some client is not selected to participate in the first fitting round and it later tries to use the full parameter exchanger to pull the partial weights, there would be a problem.
> 
> Thus, the client should only use the full parameter exchanger if it is in the first fitting round.

This implies we are sampling all clients in the first round right?

Reply:
If all clients are sampled in the first fitting round, the old code would work without any problem. But if let's say client 1 is not sampled in the first fitting round, and it is later sampled for the first time in some fitting round, then the server would have partial parameters while the client would try to use the full exchanger to pull the weights, which is a problem.

Also, because in this case the server does not necessarily have all the model parameters when a client tries to pull the weights from it for the first time, we can't follow what we did in for example `scaffold_client.py`.

Reply:
Ahh I see, thanks @yc7z! thats a much nicer property anyway (ie not having to have full participation in the first round)! I will dig into the code to understand the details more, but just initially while I wrap my head around things, I wanted to ask you: 

Why does the server model not have access to a complete set of weights to send to client who joins FL after the first round. On the first round the whole model is sent out. Subsequently, each client is going to share a subset of weights with server but I assume the server has to keep a copy of the entire model structure to map the partial weight updates to and aggregate across clients to redistribute a model the following round. Am I wrong in this assumption?